synapse/util/caches/dictionary_cache.py

```diff
@@ -96,28 +145,107 @@ def get(
         Returns:
             DictionaryEntry
         """
-        entry = self.cache.get(key, _Sentinel.sentinel)
-        if entry is not _Sentinel.sentinel:
-            if dict_keys is None:
-                return DictionaryEntry(
-                    entry.full, entry.known_absent, dict(entry.value)
-                )
+
+        if dict_keys is None:
+            # First we check if we have cached the full dict.
+            entry = self.cache.get((key, _FullCacheKey.KEY), _Sentinel.sentinel)
+            if entry is not _Sentinel.sentinel:
+                assert isinstance(entry, dict)
+                return DictionaryEntry(True, set(), entry)
+
+            # If not, check if we have cached any of dict keys.
+            all_entries = self.cache.get_multi(
+                (key,),
+                _Sentinel.sentinel,
+            )
+            if all_entries is _Sentinel.sentinel:
+                return DictionaryEntry(False, set(), {})
+
+            # If there are entries we need to unwrap the returned cache nodes
+            # and `_PerKeyValue` into the `DictionaryEntry`.
+            values = {}
+            known_absent = set()
+            for dict_key, dict_value in iterate_tree_cache_items((), all_entries):
+                dict_key = dict_key[0]
+                dict_value = dict_value.value
```

Comment:
so dict_value starts off as a `_PerKeyValue`, and we look up its `value` to get a `DV` or sentinel? in which case wtf is going on at line 177 and 179?

so wtf *is* `dict_value` ?

Reply:
OH. So I had completely forgotten this horror show, as i tried to add sanity by adding type hints to `TreeCache`, except that turned into a nightmare so didn't want to block this on that and then forgot about the initial insanity that prompted me to try and type hints.

What is going on here is that `DictionaryCache` is backed by `LruCache`, which is backed by `TreeCache`. That means that the you end up with values returned from `iterate_tree_cache_items` being `lru_cache._Node[_PerKeyValue]`.

My plan was to make `get_multi` return a generator that correctly unwrapped the returned nodes. Let me do that now